File: internal/discover/discover.go
package discover

import (
	"bufio"
	"context"
	"io"
	"log/slog"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/DeusData/codebase-memory-mcp/internal/lang"
)

// IGNORE_PATTERNS are directory names to skip during discovery.
var IGNORE_PATTERNS = map[string]bool{
	// VCS / IDE
	".git": true, ".hg": true, ".svn": true, ".worktrees": true,
	".idea": true, ".vs": true, ".vscode": true, ".eclipse": true, ".claude": true,
	// Python
	".cache": true, ".eggs": true, ".env": true, ".mypy_cache": true, ".nox": true,
	".pytest_cache": true, ".ruff_cache": true, ".tox": true, ".venv": true,
	"__pycache__": true, "env": true, "htmlcov": true, "site-packages": true, "venv": true,
	// JS/TS
	".npm": true, ".nyc_output": true, ".pnpm-store": true, ".yarn": true,
	"bower_components": true, "coverage": true, "node_modules": true,
	// JS/TS framework caches
	".next": true, ".nuxt": true, ".svelte-kit": true, ".angular": true,
	".turbo": true, ".parcel-cache": true, ".docusaurus": true, ".expo": true,
	// Build artifacts
	"dist": true, "obj": true, "Pods": true, "target": true, "temp": true, "tmp": true,
	// Build system caches
	".terraform": true, ".serverless": true,
	"bazel-bin": true, "bazel-out": true, "bazel-testlogs": true,
	// Language-specific caches
	".cargo": true, ".stack-work": true, ".dart_tool": true,
	"zig-cache": true, "zig-out": true,
	".metals": true, ".bloop": true, ".bsp": true,
	".ccls-cache": true, ".clangd": true,
	"elm-stuff": true, "_opam": true, ".cpcache": true, ".shadow-cljs": true,
	// Deploy/hosting
	".vercel": true, ".netlify": true,
	// Misc
	".qdrant_code_embeddings": true, ".tmp": true, "vendor": true,
}

// IGNORE_SUFFIXES are file suffixes that are never source files.
var IGNORE_SUFFIXES = map[string]bool{
	// Temp/compiled
	".tmp": true, "~": true, ".pyc": true, ".pyo": true,
	".o": true, ".a": true, ".so": true, ".dll": true, ".class": true,
	// Images
	".png": true, ".jpg": true, ".jpeg": true, ".gif": true,
	".ico": true, ".bmp": true, ".tiff": true, ".webp": true, ".svg": true,
	// Binaries
	".wasm": true, ".node": true, ".exe": true, ".bin": true, ".dat": true,
	// Databases
	".db": true, ".sqlite": true, ".sqlite3": true,
	// Fonts
	".woff": true, ".woff2": true, ".ttf": true, ".eot": true, ".otf": true,
}

// IndexMode controls how aggressively files are filtered during discovery.
type IndexMode string

const (
	ModeFull IndexMode = "full" // default: parse everything supported
	ModeFast IndexMode = "fast" // aggressive filtering for speed
)

// FileInfo represents a discovered source file.
type FileInfo struct {
	Path     string        // absolute path
	RelPath  string        // relative to repo root
	Language lang.Language // detected language
	Size     int64         // file size in bytes
}

// Options configures file discovery.
type Options struct {
	IgnoreFile  string    // path to .cgrignore file (optional)
	Mode        IndexMode // indexing mode (full or fast)
	MaxFileSize int64     // max file size in bytes (0 = no limit)
}

// fastIgnoreDirs are additional directories skipped in fast mode.
var fastIgnoreDirs = map[string]bool{
	"generated": true, "gen": true, "auto-generated": true,
	"fixtures": true, "testdata": true, "test_data": true,
	"__tests__": true, "__mocks__": true, "__snapshots__": true,
	"__fixtures__": true, "__test__": true,
	"docs": true, "doc": true, "documentation": true,
	"examples": true, "example": true, "samples": true, "sample": true,
	"assets": true, "static": true, "public": true, "media": true,
	"third_party": true, "thirdparty": true, "3rdparty": true, "external": true,
	"migrations": true, "seeds": true,
	"e2e": true, "integration": true,
	"locale": true, "locales": true, "i18n": true, "l10n": true,
	"scripts": true, "tools": true, "hack": true,
	// Generic dirs moved from IGNORE_PATTERNS (cause false exclusions in Go, CMake, Maven)
	"bin": true, "build": true, "out": true,
}

// fastIgnoreSuffixes are additional file extensions skipped in fast mode.
var fastIgnoreSuffixes = map[string]bool{
	// Archives
	".zip": true, ".tar": true, ".gz": true, ".bz2": true, ".xz": true,
	".rar": true, ".7z": true, ".jar": true, ".war": true, ".ear": true,
	// Media/audio/video
	".mp3": true, ".mp4": true, ".avi": true, ".mov": true, ".wav": true,
	".flac": true, ".ogg": true, ".mkv": true, ".webm": true,
	// Documents
	".pdf": true, ".doc": true, ".docx": true, ".xls": true, ".xlsx": true,
	".ppt": true, ".pptx": true, ".odt": true, ".ods": true,
	// Source maps
	".map": true,
	// Minified bundles
	".min.js": true, ".min.css": true,
	// Certificates/keys
	".pem": true, ".crt": true, ".key": true, ".cer": true, ".p12": true,
	// Serialized data
	".pb": true, ".proto": true, ".avro": true, ".parquet": true,
	// Compiled/intermediate
	".beam": true, ".elc": true, ".rlib": true,
	// Coverage/profiling
	".coverage": true, ".prof": true, ".out": true,
	// Patches
	".patch": true, ".diff": true,
}

// fastIgnoreFilenames are specific filenames skipped in fast mode.
var fastIgnoreFilenames = map[string]bool{
	"LICENSE": true, "LICENSE.txt": true, "LICENSE.md": true, "LICENSE-MIT": true, "LICENSE-APACHE": true,
	"LICENCE": true, "LICENCE.txt": true, "LICENCE.md": true,
	"CHANGELOG": true, "CHANGELOG.md": true, "CHANGES.md": true,
	"HISTORY": true, "HISTORY.md": true,
	"AUTHORS": true, "AUTHORS.md": true, "CONTRIBUTORS": true, "CONTRIBUTORS.md": true,
	"CODEOWNERS": true,
	"go.sum":     true, "yarn.lock": true, "pnpm-lock.yaml": true, "Pipfile.lock": true,
	"poetry.lock": true, "Gemfile.lock": true, "Cargo.lock": true, "mix.lock": true,
	"flake.lock": true, "pubspec.lock": true, "composer.lock": true,
	"configure": true, "Makefile.in": true, "config.guess": true, "config.sub": true,
	"package-lock.json": true,
}

// fastIgnorePatterns are suffix/contains patterns skipped in fast mode.
var fastIgnorePatterns = []string{
	".d.ts",          // TypeScript declaration files
	".bundle.",       // bundled files
	".chunk.",        // code-split chunks
	".generated.",    // generated code
	".pb.go",         // protobuf generated Go
	"_pb2.py",        // protobuf generated Python
	".pb2.py",        // protobuf generated Python (alt)
	"_grpc.pb.go",    // gRPC generated Go
	"_string.go",     // stringer generated Go
	"mock_",          // mock files prefix
	"_mock.",         // mock files suffix
	"_test_helpers.", // test helpers
	".stories.",      // Storybook stories
	".spec.",         // spec/test files
	".test.",         // test files
}

// shouldSkipDir returns true if the directory should be skipped during discovery.
func shouldSkipDir(name, rel string, extraIgnore []string, mode IndexMode) bool {
	if IGNORE_PATTERNS[name] {
		return true
	}
	if mode == ModeFast {
		// Skip all dot-directories not already in IGNORE_PATTERNS
		if strings.HasPrefix(name, ".") {
			return true
		}
		if fastIgnoreDirs[name] {
			return true
		}
	}
	for _, pattern := range extraIgnore {
		if matched, _ := filepath.Match(pattern, name); matched {
			return true
		}
		if matched, _ := filepath.Match(pattern, rel); matched {
			return true
		}
	}
	return false
}

// shouldSkipFile returns true if the file should be skipped in fast mode.
func shouldSkipFile(name, path string, size int64, opts *Options) bool {
	// File size limit (both modes)
	if opts != nil && opts.MaxFileSize > 0 && size > opts.MaxFileSize {
		return true
	}
	if opts == nil || opts.Mode != ModeFast {
		return false
	}
	// Fast-mode filename filter
	if fastIgnoreFilenames[name] {
		return true
	}
	// Fast-mode suffix filter
	for suffix := range fastIgnoreSuffixes {
		if strings.HasSuffix(path, suffix) {
			return true
		}
	}
	// Fast-mode pattern filter (contains/suffix patterns)
	for _, pattern := range fastIgnorePatterns {
		if strings.Contains(path, pattern) {
			return true
		}
	}
	return false
}

// hasIgnoredSuffix returns true if path ends with any suffix in IGNORE_SUFFIXES.
func hasIgnoredSuffix(path string) bool {
	for suffix := range IGNORE_SUFFIXES {
		if strings.HasSuffix(path, suffix) {
			return true
		}
	}
	return false
}

// Objective-C markers adapted from GitHub Linguist heuristics.yml
var objcMarkers = regexp.MustCompile(`@interface|@implementation|@protocol|@property|#import\s+[<"]|@selector|@encode|@synthesize|@dynamic`)

// MATLAB markers adapted from GitHub Linguist heuristics.yml
var matlabMarkers = regexp.MustCompile(`(?m)^\s*function\b|^\s*classdef\b|^\s*%%|^\s*%[^{]`)

// Magma markers — semicolons after end-blocks and procedure/intrinsic keywords are unique to Magma
var magmaMarkers = regexp.MustCompile(`(?m)end\s+(function|procedure|intrinsic)\s*;|\bintrinsic\s+\w+\s*\(|\bprocedure\s+\w+\s*\(|end\s+if\s*;|end\s+for\s*;|end\s+while\s*;`)

// disambiguateM determines whether a .m file is MATLAB, Objective-C, or Magma.
// Returns the detected language and true, or ("", false) if undetermined.
func disambiguateM(path string) (lang.Language, bool) {
	f, err := os.Open(path)
	if err != nil {
		return lang.MATLAB, true // default to MATLAB on error
	}
	defer f.Close()

	// Read first 4KB for heuristics
	buf := make([]byte, 4096)
	n, err := io.ReadAtLeast(f, buf, 1)
	if err != nil && n == 0 {
		return lang.MATLAB, true
	}
	head := string(buf[:n])

	if objcMarkers.MatchString(head) {
		return lang.ObjectiveC, true
	}
	// Check Magma before MATLAB — both use 'function' keyword but Magma has 'end function;'
	if magmaMarkers.MatchString(head) {
		return lang.Magma, true
	}
	if matlabMarkers.MatchString(head) {
		return lang.MATLAB, true
	}
	// Default to MATLAB (more common for .m in scientific codebases)
	return lang.MATLAB, true
}

// classifyFile checks if a file is a supported source file and returns its FileInfo.
// Returns nil if the file should be skipped.
func classifyFile(path, rel string, info os.FileInfo, opts *Options) *FileInfo {
	if hasIgnoredSuffix(path) {
		return nil
	}
	if shouldSkipFile(info.Name(), path, info.Size(), opts) {
		return nil
	}

	ext := filepath.Ext(path)

	// .m files are ambiguous: MATLAB vs Objective-C
	if ext == ".m" {
		ml, ok := disambiguateM(path)
		if ok {
			return &FileInfo{
				Path:     path,
				RelPath:  filepath.ToSlash(rel),
				Language: ml,
				Size:     info.Size(),
			}
		}
	}

	l, ok := lang.LanguageForExtension(ext)
	if !ok {
		l, ok = lang.LanguageForFilename(info.Name())
	}
	if !ok {
		return nil
	}

	if l == lang.JSON && isIgnoredJSON(info.Name()) {
		return nil
	}
	if l == lang.JSON && info.Size() > 100*1024 {
		slog.Warn("discover.skip_large_json", "path", rel, "size", info.Size())
		return nil
	}

	return &FileInfo{
		Path:     path,
		RelPath:  filepath.ToSlash(rel),
		Language: l,
		Size:     info.Size(),
	}
}

// Discover walks a repository and returns all source files.
func Discover(ctx context.Context, repoPath string, opts *Options) ([]FileInfo, error) {
	repoPath, err := filepath.Abs(repoPath)
	if err != nil {
		return nil, err
	}

	if err := ctx.Err(); err != nil {
		return nil, err
	}

	var extraIgnore []string
	if opts != nil && opts.IgnoreFile != "" {
		extraIgnore, _ = loadIgnoreFile(opts.IgnoreFile)
	} else {
		ignPath := filepath.Join(repoPath, ".cgrignore")
		extraIgnore, _ = loadIgnoreFile(ignPath)
	}

	var files []FileInfo

	mode := ModeFull
	if opts != nil && opts.Mode != "" {
		mode = opts.Mode
	}

	// Load gitignore + cbmignore matchers (nil-safe for non-git repos)
	matchers := loadIgnoreMatchers(repoPath)

	err = filepath.Walk(repoPath, func(path string, info os.FileInfo, walkErr error) error {
		if err := ctx.Err(); err != nil {
			return err
		}
		if walkErr != nil {
			return filepath.SkipDir
		}

		// Skip symlinks — prevents duplicate indexing
		if info.Mode()&os.ModeSymlink != 0 {
			return nil
		}

		rel, _ := filepath.Rel(repoPath, path)

		if info.IsDir() {
			// Hardcoded patterns first (fast O(1) map lookup)
			if shouldSkipDir(info.Name(), rel, extraIgnore, mode) {
				return filepath.SkipDir
			}
			// Then gitignore + cbmignore (skip root — library panics on base==path)
			if rel != "." && matchers.shouldIgnore(path, true) {
				slog.Debug("discover.gitignore_skip", "dir", rel)
				return filepath.SkipDir
			}
			return nil
		}

		// Gitignore + cbmignore check for files
		if matchers.shouldIgnore(path, false) {
			return nil
		}

		if fi := classifyFile(path, rel, info, opts); fi != nil {
			files = append(files, *fi)
		}
		return nil
	})

	return files, err
}

// ignoredJSONFiles are JSON filenames to skip (tool configs, lock files, specs).
var ignoredJSONFiles = map[string]bool{
	"package.json":       true,
	"package-lock.json":  true,
	"tsconfig.json":      true,
	"jsconfig.json":      true,
	"composer.json":      true,
	"composer.lock":      true,
	"yarn.lock":          true,
	"openapi.json":       true,
	"swagger.json":       true,
	"jest.config.json":   true,
	".eslintrc.json":     true,
	".prettierrc.json":   true,
	".babelrc.json":      true,
	"tslint.json":        true,
	"angular.json":       true,
	"firebase.json":      true,
	"renovate.json":      true,
	"lerna.json":         true,
	"turbo.json":         true,
	".stylelintrc.json":  true,
	"pnpm-lock.json":     true,
	"deno.json":          true,
	"biome.json":         true,
	"devcontainer.json":  true,
	".devcontainer.json": true,
	"launch.json":        true,
	"settings.json":      true,
	"extensions.json":    true,
	"tasks.json":         true,
}

func isIgnoredJSON(name string) bool {
	return ignoredJSONFiles[name]
}

func loadIgnoreFile(path string) ([]string, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var patterns []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" && !strings.HasPrefix(line, "#") {
			patterns = append(patterns, line)
		}
	}
	return patterns, scanner.Err()
}

File: src/discover/discover.c
/*
 * discover.c — Recursive directory walk with filtering.
 *
 * Walks a repository directory tree, applying:
 *   1. Hardcoded directory skip patterns (60+ dirs like .git, node_modules)
 *   2. Hardcoded suffix filters (.pyc, .png, .wasm, etc.)
 *   3. Fast-mode additional filters (docs, examples, lock files, etc.)
 *   4. Gitignore-style pattern matching
 *   5. Language detection for accepted files
 */
#include "discover/discover.h"
#include "cbm.h" // CBMLanguage, CBM_LANG_COUNT, CBM_LANG_JSON

#include "foundation/constants.h"
#include "foundation/compat_fs.h"
#include <stdint.h> // int64_t
#include <stdio.h>
#include <stdlib.h>
#include <string.h> // strdup
#include <sys/stat.h>

/* ── Hardcoded always-skip directories ───────────────────────────── */

static const char *ALWAYS_SKIP_DIRS[] = {
    /* VCS */
    ".git", ".hg", ".svn", ".worktrees",
    /* IDE */
    ".idea", ".vs", ".vscode", ".eclipse", ".claude",
    /* Python */
    ".cache", ".eggs", ".env", ".mypy_cache", ".nox", ".pytest_cache", ".ruff_cache", ".tox",
    ".venv", "__pycache__", "env", "htmlcov", "site-packages", "venv",
    /* JS/TS */
    ".npm", ".nyc_output", ".pnpm-store", ".yarn", "bower_components", "coverage", "node_modules",
    ".next", ".nuxt", ".svelte-kit", ".angular", ".turbo", ".parcel-cache", ".docusaurus", ".expo",
    /* Build artifacts */
    "dist", "obj", "Pods", "target", "temp", "tmp", ".terraform", ".serverless", "bazel-bin",
    "bazel-out", "bazel-testlogs",
    /* Language caches */
    ".cargo", ".stack-work", ".dart_tool", "zig-cache", "zig-out", ".metals", ".bloop", ".bsp",
    ".ccls-cache", ".clangd", "elm-stuff", "_opam", ".cpcache", ".shadow-cljs",
    /* Deploy */
    ".vercel", ".netlify",
    /* Misc */
    ".qdrant_code_embeddings", ".tmp", "vendor", NULL};

static const char *FAST_SKIP_DIRS[] = {
    "generated", "gen",           "auto-generated", "fixtures",     "testdata",    "test_data",
    "__tests__", "__mocks__",     "__snapshots__",  "__fixtures__", "__test__",    "docs",
    "doc",       "documentation", "examples",       "example",      "samples",     "sample",
    "assets",    "static",        "public",         "media",        "third_party", "thirdparty",
    "3rdparty",  "external",      "migrations",     "seeds",        "e2e",         "integration",
    "locale",    "locales",       "i18n",           "l10n",         "scripts",     "tools",
    "hack",      "bin",           "build",          "out",          NULL};

/* ── Ignored suffixes ────────────────────────────────────────────── */

static const char *ALWAYS_IGNORED_SUFFIXES[] = {
    ".tmp",    "~",        ".pyc",  ".pyo",   ".o",   ".a",   ".so",  ".dll",
    ".class",  ".png",     ".jpg",  ".jpeg",  ".gif", ".ico", ".bmp", ".tiff",
    ".webp",   ".svg",     ".wasm", ".node",  ".exe", ".bin", ".dat", ".db",
    ".sqlite", ".sqlite3", ".woff", ".woff2", ".ttf", ".eot", ".otf", NULL};

static const char *FAST_IGNORED_SUFFIXES[] = {
    ".zip", ".tar",  ".gz",       ".bz2",  ".xz",  ".rar",    ".7z",      ".jar",
    ".war", ".ear",  ".mp3",      ".mp4",  ".avi", ".mov",    ".wav",     ".flac",
    ".ogg", ".mkv",  ".webm",     ".pdf",  ".doc", ".docx",   ".xls",     ".xlsx",
    ".ppt", ".pptx", ".odt",      ".ods",  ".map", ".min.js", ".min.css", ".pem",
    ".crt", ".key",  ".cer",      ".p12",  ".pb",  ".avro",   ".parquet", ".beam",
    ".elc", ".rlib", ".coverage", ".prof", ".out", ".patch",  ".diff",    NULL};

/* ── Fast-mode skip filenames ────────────────────────────────────── */

static const char *FAST_SKIP_FILENAMES[] = {
    "LICENSE",        "LICENSE.txt",     "LICENSE.md",   "LICENSE-MIT",   "LICENSE-APACHE",
    "LICENCE",        "LICENCE.txt",     "LICENCE.md",   "CHANGELOG",     "CHANGELOG.md",
    "CHANGES.md",     "HISTORY",         "HISTORY.md",   "AUTHORS",       "AUTHORS.md",
    "CONTRIBUTORS",   "CONTRIBUTORS.md", "CODEOWNERS",   "go.sum",        "yarn.lock",
    "pnpm-lock.yaml", "Pipfile.lock",    "poetry.lock",  "Gemfile.lock",  "Cargo.lock",
    "mix.lock",       "flake.lock",      "pubspec.lock", "composer.lock", "package-lock.json",
    "configure",      "Makefile.in",     "config.guess", "config.sub",    NULL};

/* ── Fast-mode substring patterns ────────────────────────────────── */

static const char *FAST_PATTERNS[] = {".d.ts",      ".bundle.", ".chunk.", ".generated.",
                                      ".pb.go",     "_pb2.py",  ".pb2.py", "_grpc.pb.go",
                                      "_string.go", "mock_",    "_mock.",  "_test_helpers.",
                                      ".stories.",  ".spec.",   ".test.",  NULL};

/* ── Ignored JSON filenames ──────────────────────────────────────── */

static const char *IGNORED_JSON_FILES[] = {
    "package.json",       "package-lock.json", "tsconfig.json",
    "jsconfig.json",      "composer.json",     "composer.lock",
    "yarn.lock",          "openapi.json",      "swagger.json",
    "jest.config.json",   ".eslintrc.json",    ".prettierrc.json",
    ".babelrc.json",      "tslint.json",       "angular.json",
    "firebase.json",      "renovate.json",     "lerna.json",
    "turbo.json",         ".stylelintrc.json", "pnpm-lock.json",
    "deno.json",          "biome.json",        "devcontainer.json",
    ".devcontainer.json", "launch.json",       "settings.json",
    "extensions.json",    "tasks.json",        NULL};

/* ── Helper: check if string is in NULL-terminated array ─────────── */

static bool str_in_list(const char *s, const char *const *list) {
    for (int i = 0; list[i]; i++) {
        if (strcmp(s, list[i]) == 0) {
            return true;
        }
    }
    return false;
}

/* ── Helper: check if string ends with suffix ────────────────────── */

static bool ends_with(const char *s, const char *suffix) {
    size_t slen = strlen(s);
    size_t sufflen = strlen(suffix);
    if (sufflen > slen) {
        return false;
    }
    return strcmp(s + slen - sufflen, suffix) == 0;
}

/* ── Helper: check if string contains substring ──────────────────── */

static bool str_contains(const char *s, const char *sub) {
    return strstr(s, sub) != NULL;
}

/* ── Public filter functions ─────────────────────────────────────── */

bool cbm_should_skip_dir(const char *dirname, cbm_index_mode_t mode) {
    if (!dirname) {
        return false;
    }

    if (str_in_list(dirname, ALWAYS_SKIP_DIRS)) {
        return true;
    }

    /* Fast discovery applies to both MODERATE and FAST — only FULL keeps everything. */
    if (mode != CBM_MODE_FULL) {
        if (str_in_list(dirname, FAST_SKIP_DIRS)) {
            return true;
        }
    }

    return false;
}

bool cbm_has_ignored_suffix(const char *filename, cbm_index_mode_t mode) {
    if (!filename) {
        return false;
    }

    for (int i = 0; ALWAYS_IGNORED_SUFFIXES[i]; i++) {
        if (ends_with(filename, ALWAYS_IGNORED_SUFFIXES[i])) {
            return true;
        }
    }

    if (mode != CBM_MODE_FULL) {
        for (int i = 0; FAST_IGNORED_SUFFIXES[i]; i++) {
            if (ends_with(filename, FAST_IGNORED_SUFFIXES[i])) {
                return true;
            }
        }
    }

    return false;
}

bool cbm_should_skip_filename(const char *filename, cbm_index_mode_t mode) {
    if (!filename) {
        return false;
    }

    if (mode != CBM_MODE_FULL) {
        if (str_in_list(filename, FAST_SKIP_FILENAMES)) {
            return true;
        }
    }

    return false;
}

bool cbm_matches_fast_pattern(const char *filename, cbm_index_mode_t mode) {
    if (!filename || mode == CBM_MODE_FULL) {
        return false;
    }

    for (int i = 0; FAST_PATTERNS[i]; i++) {
        if (str_contains(filename, FAST_PATTERNS[i])) {
            return true;
        }
    }

    return false;
}

/* ── Dynamic file list ───────────────────────────────────────────── */

typedef struct {
    cbm_file_info_t *files;
    int count;
    int capacity;
} file_list_t;

static void fl_add(file_list_t *fl, const char *abs_path, const char *rel_path, CBMLanguage lang,
                   int64_t size) {
    if (fl->count >= fl->capacity) {
        int new_cap = fl->capacity ? fl->capacity * PAIR_LEN : CBM_SZ_256;
        cbm_file_info_t *new_files = realloc(fl->files, new_cap * sizeof(cbm_file_info_t));
        if (!new_files) {
            return;
        }
        fl->files = new_files;
        fl->capacity = new_cap;
    }

    cbm_file_info_t *fi = &fl->files[fl->count++];
    fi->path = strdup(abs_path);
    fi->rel_path = strdup(rel_path);
    fi->language = lang;
    fi->size = size;
}

/* ── Recursive walk ──────────────────────────────────────────────── */

/* Check if a directory entry should be skipped (hardcoded dirs + gitignore). */
static bool should_skip_directory(const char *entry_name, const char *rel_path,
                                  const cbm_discover_opts_t *opts, const cbm_gitignore_t *gitignore,
                                  const cbm_gitignore_t *cbmignore) {
    if (cbm_should_skip_dir(entry_name, opts ? opts->mode : CBM_MODE_FULL)) {
        return true;
    }
    if (gitignore && cbm_gitignore_matches(gitignore, rel_path, true)) {
        return true;
    }
    if (cbmignore && cbm_gitignore_matches(cbmignore, rel_path, true)) {
        return true;
    }
    return false;
}

/* Check if a regular file should be skipped (filters + gitignore + size). */
static bool should_skip_file(const char *entry_name, const char *rel_path,
                             const cbm_discover_opts_t *opts, const cbm_gitignore_t *gitignore,
                             const cbm_gitignore_t *cbmignore, off_t file_size) {
    cbm_index_mode_t mode = opts ? opts->mode : CBM_MODE_FULL;
    if (cbm_has_ignored_suffix(entry_name, mode)) {
        return true;
    }
    if (cbm_should_skip_filename(entry_name, mode)) {
        return true;
    }
    if (cbm_matches_fast_pattern(entry_name, mode)) {
        return true;
    }
    if (gitignore && cbm_gitignore_matches(gitignore, rel_path, false)) {
        return true;
    }
    if (cbmignore && cbm_gitignore_matches(cbmignore, rel_path, false)) {
        return true;
    }
    if (opts && opts->max_file_size > 0 && file_size > opts->max_file_size) {
        return true;
    }
    return false;
}

/* Detect language for a file, handling .m disambiguation and JSON filtering. */
static CBMLanguage detect_file_language(const char *entry_name, const char *abs_path) {
    CBMLanguage lang = cbm_language_for_filename(entry_name);
    if (lang == CBM_LANG_COUNT) {
        return CBM_LANG_COUNT;
    }
    /* Special: .m files need content-based disambiguation */
    const char *dot = strrchr(entry_name, '.');
    if (dot && strcmp(dot, ".m") == 0) {
        lang = cbm_disambiguate_m(abs_path);
    }
    /* Check ignored JSON files */
    if (lang == CBM_LANG_JSON && str_in_list(entry_name, IGNORED_JSON_FILES)) {
        return CBM_LANG_COUNT;
    }
    return lang;
}

/* Stat a path, skipping symlinks. Returns 0 on success, -1 to skip. */
static int safe_stat(const char *abs_path, struct stat *st) {
#ifdef _WIN32
    if (stat(abs_path, st) != 0) {
        return CBM_NOT_FOUND;
    }
#else
    if (lstat(abs_path, st) != 0) {
        return CBM_NOT_FOUND;
    }
    if (S_ISLNK(st->st_mode)) {
        return CBM_NOT_FOUND;
    }
#endif
    return 0;
}

/* Process a single regular file entry during directory walk. */
static void walk_dir_process_file(const char *abs_path, const char *rel_path, const char *name,
                                  const cbm_discover_opts_t *opts, const cbm_gitignore_t *gitignore,
                                  const cbm_gitignore_t *cbmignore, off_t size, file_list_t *out) {
    if (should_skip_file(name, rel_path, opts, gitignore, cbmignore, size)) {
        return;
    }
    CBMLanguage lang = detect_file_language(name, abs_path);
    if (lang == CBM_LANG_COUNT) {
        return;
    }
    fl_add(out, abs_path, rel_path, lang, size);
}

typedef struct {
    char dir[CBM_SZ_4K];
    char prefix[CBM_SZ_4K];
} walk_frame_t;
#define WALK_STACK_CAP 512
/* Build abs/rel paths and process one directory entry. */
static void walk_dir_process_entry(cbm_dirent_t *entry, const walk_frame_t *frame,
                                   const cbm_discover_opts_t *opts,
                                   const cbm_gitignore_t *gitignore,
                                   const cbm_gitignore_t *cbmignore, walk_frame_t *stack, int *top,
                                   file_list_t *out) {
    char abs_path[CBM_SZ_4K];
    char rel_path[CBM_SZ_4K];
    snprintf(abs_path, sizeof(abs_path), "%s/%s", frame->dir, entry->name);
    if (frame->prefix[0] != '\0') {
        snprintf(rel_path, sizeof(rel_path), "%s/%s", frame->prefix, entry->name);
    } else {
        snprintf(rel_path, sizeof(rel_path), "%s", entry->name);
    }

    struct stat st;
    if (safe_stat(abs_path, &st) != 0) {
        return;
    }

    if (S_ISDIR(st.st_mode)) {
        if (!should_skip_directory(entry->name, rel_path, opts, gitignore, cbmignore)) {
            if (*top < WALK_STACK_CAP) {
                snprintf(stack[*top].dir, CBM_SZ_4K, "%s", abs_path);
                snprintf(stack[*top].prefix, CBM_SZ_4K, "%s", rel_path);
                (*top)++;
            }
        }
    } else if (S_ISREG(st.st_mode)) {
        walk_dir_process_file(abs_path, rel_path, entry->name, opts, gitignore, cbmignore,
                              st.st_size, out);
    }
}

static void walk_dir(const char *dir_path, const char *rel_prefix, const cbm_discover_opts_t *opts,
                     const cbm_gitignore_t *gitignore, const cbm_gitignore_t *cbmignore,
                     file_list_t *out) {
    walk_frame_t *stack = calloc(WALK_STACK_CAP, sizeof(walk_frame_t));
    if (!stack) {
        return;
    }
    int top = 0;
    snprintf(stack[top].dir, CBM_SZ_4K, "%s", dir_path);
    snprintf(stack[top].prefix, CBM_SZ_4K, "%s", rel_prefix);
    top++;

    while (top > 0) {
        walk_frame_t frame = stack[--top];
        cbm_dir_t *d = cbm_opendir(frame.dir);
        if (!d) {
            continue;
        }

        cbm_dirent_t *entry;
        while ((entry = cbm_readdir(d)) != NULL) {
            walk_dir_process_entry(entry, &frame, opts, gitignore, cbmignore, stack, &top, out);
        }
        cbm_closedir(d);
    }
    free(stack);
}

/* ── Public API ──────────────────────────────────────────────────── */

int cbm_discover(const char *repo_path, const cbm_discover_opts_t *opts, cbm_file_info_t **out,
                 int *count) {
    if (!repo_path || !out || !count) {
        return CBM_NOT_FOUND;
    }

    *out = NULL;
    *count = 0;

    /* Verify directory exists */
    struct stat st;
    if (stat(repo_path, &st) != 0 || !S_ISDIR(st.st_mode)) {
        return CBM_NOT_FOUND;
    }

    /* Load gitignore if .git directory exists */
    cbm_gitignore_t *gitignore = NULL;
    char gi_path[CBM_SZ_4K];
    snprintf(gi_path, sizeof(gi_path), "%s/.git", repo_path);
    struct stat gi_stat;
    if (stat(gi_path, &gi_stat) == 0 && S_ISDIR(gi_stat.st_mode)) {
        snprintf(gi_path, sizeof(gi_path), "%s/.gitignore", repo_path);
        gitignore = cbm_gitignore_load(gi_path);
    }

    /* Load cbmignore if specified or exists at repo root */
    cbm_gitignore_t *cbmignore = NULL;
    if (opts && opts->ignore_file) {
        cbmignore = cbm_gitignore_load(opts->ignore_file);
    } else {
        snprintf(gi_path, sizeof(gi_path), "%s/.cbmignore", repo_path);
        cbmignore = cbm_gitignore_load(gi_path);
    }

    /* Walk */
    file_list_t fl = {0};
    walk_dir(repo_path, "", opts, gitignore, cbmignore, &fl);

    /* Cleanup */
    cbm_gitignore_free(gitignore);
    cbm_gitignore_free(cbmignore);

    *out = fl.files;
    *count = fl.count;
    return 0;
}

void cbm_discover_free(cbm_file_info_t *files, int count) {
    if (!files) {
        return;
    }
    for (int i = 0; i < count; i++) {
        free(files[i].path);
        free(files[i].rel_path);
    }
    free(files);
}
